### PR TITLE
add me-tokens to dropdown

### PR DESCRIPTION
**Pull request by Seroxdesign**
## Overview

**What features/fixes does this PR include?**

Integrates Metokens into the MetaOS custom blocks

**Please provide the GitHub issue number**

## Follow up Improvement Ideas

- [ ] please list any improvement/ideas

## Implementation

**Describe technical (nontrivial / non-obvious) parts of your code**

**Side effects**

## Assets

[Include screenshots/videos if it makes reviewing easier.]

### CODE REVIEW

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+    };
+    getInAndOutTokenData();
+  }, [metokenAddress, collateral]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  useEffect(() => {
+    if (!amount || !metokenAddress) return;
+    handlePreview();
+  }, [transactionType, amount, metokenAddress]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const handleSpendMeTokens = async () => {
```
Comment (alalonde) [inline packages/web/components/Player/Section/PlayerMeToken.tsx:118]: All these functions should either be wrapped in `useCallback` or moved outside the render function. As-is they are recreated on every single rendering pass

File: packages/web/utils/orbis.d.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+declare module '@orbisclub/orbis-sdk';
```
Comment (alalonde) [inline packages/web/utils/orbis.d.ts:1]: Can you put this with the other external types in `packages/@types`?

File: packages/web/utils/meTokens.tsx
Comment (alalonde) [inline packages/web/utils/meTokens.tsx:1]: If this doesn't contain components it should be `.ts`

File: packages/web/utils/meTokens.tsx [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+import { BigNumber, Contract, ethers } from 'ethers';
+
+export const metokenRegistry = '0x8b91FcF2230ab04A46e2D83aaF062EC1B5AAAa5c';
+export const metokenFactory = '0x8D4ee3599aF814bF3Aa884c161f0dE81d9e97225';
+export const hubFacet = '0x4555cf6E984186F6C0dfeba1A26764b21553B39f';
+export const foundryFacet = '0xA56AAF637b057a5EDf7b7252D0B7280042E71335';
+export const metokenDiamond = '0x0B4ec400e8D10218D0869a5b0036eA4BCf92d905';
+export const nullMeToken = '0x0000000000000000000000000000000000000000';
+
+const COINGECKO_API_URL = 'https://api.coingecko.com/api/v3/coins/';
```
Comment (alalonde) [inline packages/web/utils/meTokens.tsx:17]: I believe we have this constant defined elsewhere. If so, use that one

File: packages/web/utils/meTokens.tsx [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  type: string,
+) => {
+  if (!amount || !meTokenAddress) return BigNumber.from(0);
+  const meTokenFoundry = await new Contract(
+    metokenDiamond,
+    FoundryFacetAbi,
+    mainnetProvider.getSigner(senderAddress),
+  );
+  if (type === 'mint') {
+    const tx = await meTokenFoundry
```
Comment (alalonde) [inline packages/web/utils/meTokens.tsx:128]: probably shouldn't be using `await` and `.then`

File: packages/web/utils/meTokens.tsx [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+export const metokenDiamond = '0x0B4ec400e8D10218D0869a5b0036eA4BCf92d905';
+export const nullMeToken = '0x0000000000000000000000000000000000000000';
+
+const COINGECKO_API_URL = 'https://api.coingecko.com/api/v3/coins/';
+const COINQUERY = '?localization=false&tickers=true&market_data=true';
+export const DAIADDRESS = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
+
+export const getCollateralData = async (collateralTokenAddress: string) => {
+  let id;
+  switch (collateralTokenAddress) {
```
Comment (alalonde) [inline packages/web/utils/meTokens.tsx:23]: what is the point of this switch statement?

File: packages/web/contracts/FoundryFacet.abi.ts
Comment (alalonde) [inline packages/web/contracts/FoundryFacet.abi.ts:1]: @dysbulic What is the best practice for including these contract definitions? Seems like we should be pulling them in from elsewhere if possible

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+        setMeTokenData(res);
+      });
+    };
+    getInAndOutTokenData();
+  }, [metokenAddress, collateral]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  useEffect(() => {
+    if (!amount || !metokenAddress) return;
+    handlePreview();
+  }, [transactionType, amount, metokenAddress]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (alalonde) [inline packages/web/components/Player/Section/PlayerMeToken.tsx:116]: Ditto, handlePreview itself has a bunch of dependencies

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+    } else {
+      handleSetAmount(spendAmount);
+    }
+  };
+
+  const handleSwitchToMainnet = async () => {
+    await switchChainOnMetaMask('0x1');
+  };
+
+  const roundNumber = (number: string) =>
```
Comment (alalonde) [inline packages/web/components/Player/Section/PlayerMeToken.tsx:288]: either use an existing function in `utils` or `packages/utils` or move these utility functions there

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+    if (transactionType === 'mint') {
+      changeTransactionType();
+      handleSetAmount(spendAmount);
+    } else {
+      handleSetAmount(spendAmount);
+    }
+  };
+
+  const handleSwitchToMainnet = async () => {
+    await switchChainOnMetaMask('0x1');
```
Comment (alalonde) [inline packages/web/components/Player/Section/PlayerMeToken.tsx:285]: what about other wallet providers?

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+                <MetaButton
+                  mx="auto"
+                  mt="1rem"
+                  onClick={handleSpendMeTokens}
+                  disabled={loading}
+                >
+                  {loading ? 'Loading..' : `Spend ${symbol}`}
+                </MetaButton>
+              ) : (
+                <MetaButton mx="auto" mt="1rem" onClick={handleSwitchToMainnet}>
```
Comment (alalonde) [inline packages/web/components/Player/Section/PlayerMeToken.tsx:540]: I think there is an existing component that does this. See `SwitchNetworkButton`

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+  useEffect(() => {
+    if (!player) return;
+    const getTokenByOwner = async () => {
+      await getMeTokenFor(player?.ethereumAddress).then((r) => {
+        setMeTokenAddress(r === nullMeToken ? 'Create meToken' : r);
+      });
+    };
+
+    getTokenByOwner();
+  }, [player?.ethereumAddress]); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (alalonde) [inline packages/web/components/Player/Section/PlayerMeToken.tsx:591]: Please explain why  dependencies are omitted in a comment or fix. Same with the one below

**Review by alalonde — CHANGES_REQUESTED** [review]
Thanks for putting this together!

My main concern is what seems to be improper `useEffect`s throughout.

File: packages/web/utils/meTokens.tsx (hunk 5 above)
Comment (Seroxdesign) [reply packages/web/utils/meTokens.tsx:23]: There will be more collateral tokens, we match them with their ID to query coingecko api

File: packages/web/components/Player/Section/PlayerMeToken.tsx (hunk 7 above)
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx:288]: Moved to utils/mathhelpers

File: packages/web/components/Player/Section/PlayerMeToken.tsx (hunk 8 above)
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx:285]: Used SwitchNetworkButton (as mentioned in comment below)

File: packages/web/components/Player/Section/PlayerMeToken.tsx (hunk 9 above)
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx:540]: Used SwitchNetworkButton

File: packages/web/utils/meTokens.tsx (hunk 4 above)
Comment (Seroxdesign) [reply packages/web/utils/meTokens.tsx]: Resolved by using try/catch blocks, example:

```
   try {
      const tx = await meTokenFoundry
      .calculateMeTokensMinted(meTokenAddress, amount)
      return tx;
    } catch (e) {
      throw e;
    }
 
```

File: packages/web/components/Player/Section/PlayerMeToken.tsx [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+    const getInAndOutTokenData = async () => {
+      await getErc20TokenData(collateral, owner).then((res) => {
+        setCollateralTokenData(res);
+      });
+      await getErc20TokenData(metokenAddress, owner).then((res) => {
+        setMeTokenData(res);
+      });
+    };
+    getInAndOutTokenData();
+  }, [metokenAddress, collateral]); // eslint-disable-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx:111]: Added all the deps in array

File: packages/web/components/Player/Section/PlayerMeToken.tsx (hunk 6 above)
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx:116]: Moved this to a useCallback

File: packages/web/components/Player/Section/PlayerMeToken.tsx (hunk 10 above)
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx:591]: Added all deps to dependency array

File: packages/web/components/Player/Section/PlayerMeToken.tsx (hunk 1 above)
Comment (Seroxdesign) [reply packages/web/components/Player/Section/PlayerMeToken.tsx]: Wrapped all functions in useCallback, moved utils out to utils/numbers

File: packages/web/contracts/FoundryFacet.abi.ts
Comment (dysbulic) [reply packages/web/contracts/FoundryFacet.abi.ts:1]: They're coming from Etherscan, so we have to include them in our repo somewhere.

**Review by alalonde — APPROVED** [review]

**Comment by alalonde** [comment]
I would like to see someone from meTokens view [this deployment](https://frontend-pr-1541-mjhnbmqqna-uk.a.run.app/) before we merge this.

Also this needs a rebase still.

**Comment by Seroxdesign** [comment]
> I would like to see someone from meTokens view [this deployment](https://frontend-pr-1541-mjhnbmqqna-uk.a.run.app/) before we merge this.
> 
> Also this needs a rebase still.

I've rebased, waiting on an approval comment from Cbob @metokens

**Pull request closed without merging**